File: src/main/java/ch/obermuhlner/astro/gradient/operation/AbstractSimplePixelImageOperation.java
package ch.obermuhlner.astro.gradient.operation;

import ch.obermuhlner.astro.image.DoubleImage;
import ch.obermuhlner.astro.image.color.ColorModel;

public abstract class AbstractSimplePixelImageOperation implements ImageOperation {

  @Override
  public DoubleImage operation(DoubleImage image1, DoubleImage image2, DoubleImage result) {
    double[] pixel1 = new double[3];
    double[] pixel2 = new double[3];
    double[] pixelResult = new double[3];
    ColorModel colorModel = result.getColorModel();

    for (int y = 0; y < image1.getHeight(); y++) {
      for (int x = 0; x < image1.getWidth(); x++) {
        image1.getPixel(x, y, colorModel, pixel1);
        image2.getPixel(x, y, colorModel, pixel2);
        result.setPixel(x, y, colorModel, pixelOperation(pixel1, pixel2, x, y, pixelResult));
      }
    }

    return result;
  }

  protected abstract double[] pixelOperation(double[] pixel1, double[] pixel2, int x, int y, double[] result);
}

File: src/main/java/ch/obermuhlner/astro/gradient/operation/AbstractSimplePixelImageOperation.kt
package ch.obermuhlner.astro.gradient.operation

import ch.obermuhlner.astro.image.DoubleImage
import ch.obermuhlner.astro.image.color.ColorModel

abstract class AbstractSimplePixelImageOperation : ImageOperation {
    override fun operation(image1: DoubleImage, image2: DoubleImage, result: DoubleImage): DoubleImage {
        val pixel1: DoubleArray = DoubleArray(3)
        val pixel2: DoubleArray = DoubleArray(3)
        val pixelResult: DoubleArray = DoubleArray(3)
        val colorModel: ColorModel = result.colorModel
        for (y in 0 until image1.height) {
            for (x in 0 until image1.width) {
                image1.getPixel(x, y, colorModel, pixel1)
                image2.getPixel(x, y, colorModel, pixel2)
                result.setPixel(x, y, colorModel, pixelOperation(pixel1, pixel2, x, y, pixelResult))
            }
        }
        return result
    }

    protected abstract fun pixelOperation(pixel1: DoubleArray, pixel2: DoubleArray, x: Int, y: Int, result: DoubleArray): DoubleArray
}